opm/autodiff/AquiferFetkovich.hpp

```diff
@@ -21,367 +21,214 @@ along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+              {
+                Base::faceArea_connected_.at(idx) =  Base::getFaceArea(faceCells, ugrid, faceIdx, idx, connection);
+                denom_face_areas += ( connection.influx_multiplier.at(idx) * Base::faceArea_connected_.at(idx) );
+              }
+            }
+            auto cellCenter = grid.getCellCenter(Base::cell_idx_.at(idx));
+            Base::cell_depth_.at(idx) = cellCenter[2];
+          }
 
-        inline void updateCellDensity(const int idx, const IntensiveQuantities& intQuants)
-        {
-          const auto& fs = intQuants.fluidState();
-          rhow_.at(idx) = fs.density(waterPhaseIdx);
+          const double eps_sqrt = std::sqrt(std::numeric_limits<double>::epsilon());
+          for (size_t idx = 0; idx < Base::cell_idx_.size(); ++idx)
+          {
+            Base::alphai_.at(idx) = (denom_face_areas < eps_sqrt)? // Prevent no connection NaNs due to division by zero
+            0.
+            : ( connection.influx_multiplier.at(idx) * Base::faceArea_connected_.at(idx) )/denom_face_areas;
+          }
         }
 
         inline Scalar dpai(int idx)
         {
-          Scalar dp = aquifer_pressure_ + rhow_.at(idx).value()*gravity_()*(cell_depth_.at(idx) - aqufetp_data_.d0) - pressure_current_.at(idx).value() ;
+          Scalar dp = aquifer_pressure_ + Base::rhow_.at(idx).value()*Base::gravity_()*(Base::cell_depth_.at(idx) - aqufetp_data_.d0) - Base::pressure_current_.at(idx).value() ;
           return dp;
         }
         // This function implements Eq 5.12 of the EclipseTechnicalDescription
         inline Scalar aquiferPressure()
         {
-          Scalar Flux = W_flux_.value();
-          Scalar pa_ = pa0_ - Flux / ( aqufetp_data_.C_t * aqufetp_data_.V0 );
+          Scalar Flux = Base::W_flux_.value();
+          Scalar pa_ = Base::pa0_ - Flux / ( aqufetp_data_.C_t * aqufetp_data_.V0 );
           return pa_;
         }
         // This function implements Eq 5.14 of the EclipseTechnicalDescription
         inline void calculateInflowRate(int idx, const Simulator& simulator)
         {
-          Tc_ = ( aqufetp_data_.C_t * aqufetp_data_.V0 ) / aqufetp_data_.J ;
-          Scalar td_Tc_ = simulator.timeStepSize() / Tc_ ;
+          Base::Tc_ = ( aqufetp_data_.C_t * aqufetp_data_.V0 ) / aqufetp_data_.J ;
+          Scalar td_Tc_ = simulator.timeStepSize() / Base::Tc_ ;
           Scalar exp_ = (1 - exp(-td_Tc_)) / td_Tc_;
-          Qai_.at(idx) = alphai_.at(idx) * aqufetp_data_.J * dpai(idx) * exp_;
+          Base::Qai_.at(idx) = Base::alphai_.at(idx) * aqufetp_data_.J * dpai(idx) * exp_;
         }
 
-        template<class faceCellType, class ugridType>
-        inline const double getFaceArea(const faceCellType& faceCells, const ugridType& ugrid,
-                                        const int faceIdx, const int idx,
-                                        const Aquancon::AquanconOutput& connection) const
+        inline void calculateAquiferCondition()

```

Comment:
Please also check whether this function can go to the base class.

Reply:
This cannot as it uses the input deck data. Therefore it is specific to the object. Maybe, this can changed in the future but for now I will leave it like this.